tests/server.py

```diff
@@ -1029,18 +1029,15 @@ def setup_test_homeserver(
 
     # Create the database before we actually try and connect to it, based off
     # the template database we generate in setupdb()
-    if isinstance(db_engine, PostgresEngine):
-        import psycopg2.extensions
-
+    if USE_POSTGRES_FOR_TESTS:
```

Comment:
This change is weirdly so that mypy treats `db_engine` as a `BaseDatabaseEngine` and not a `PostgresEngine` or it dislikes that `db_conn` is a `Connection` (protocol) instead of  `psycopg2.extensions.connection`.

Reply:
Probably related to https://github.com/matrix-org/synapse/blob/4f1840a88ad3a93244fc23149c56245704eab824/synapse/storage/engines/postgres.py#L34-L36 which was my way of trying to teach mypy which connection and cursor type to expect when we're using the PostgresEngine.

Reply:
(Although if we had to do the type-assert on line -1042 then it clearly wasn't working.)

Reply:
Right -- from those mypy is able to deduce that the `PostgresEngine` creates a `psycopg2.extensions.connection`; *but* the API expects a `Connection` protocol, and it doesn't seem to understand that `psycopg2.extensions.connection` matches `Connection`?

I wonder if we should make `LoggingTransaction` etc. generic and be passing these through so that the type checking knows the *real* types of things? I guess you still need a generic protocol type though for use with the `BaseDatabaseEngine` class.

Reply:
> it doesn't seem to understand that `psycopg2.extensions.connection` matches `Connection`?

That's a surprise to me:

```
2023-11-09 21:04:54 ✔  $ cat temp.py
import psycopg2.extensions
import typing
import synapse.storage.types

pg_conn: psycopg2.extensions.connection
generic_conn: synapse.storage.types.Connection

generic_conn = pg_conn


dmr on titan in synapse on  develop is 📦 v1.96.0rc1 via 🐍 v3.11.6 (matrix-synapse-py3.11) via 🦀 v1.68.0 
2023-11-09 21:05:01 ✔  $ mypy temp.py 
Success: no issues found in 1 source file
```

Reply:
```
tests/server.py:1085: error: Argument 1 to "attempt_to_set_autocommit" of "PostgresEngine" has incompatible type "Connection"; expected "connection"  [arg-type]
```

More than happy to do another PR changing it back though?

Reply:
:shrug: No strong opinions here!